File: ios/ColdOpenCore/Sources/ColdOpenCore/FavoritesStore.swift
import Foundation

/// Shared favorite-quote storage — written by the container app (the
/// quote browser's favorite toggle, see `ContentView.swift`) and read by
/// the widget (to exclude favorites from the randomize/auto-refresh
/// pool, see `ColdOpenWidget.swift`), via the same App Group suite
/// `QuoteSelector`'s "already shown" persistence already uses.
///
/// The `UserDefaults` instance is injected rather than looked up
/// internally via `AppGroup.identifier`, so this stays testable without
/// a real App Group — callers pass `UserDefaults(suiteName:
/// AppGroup.identifier)` (or `nil`, which this degrades gracefully
/// from, same as the rest of this App Group's readers).
public struct FavoritesStore {
    private static let key = "favoriteQuoteTexts"
    private let defaults: UserDefaults?

    public init(defaults: UserDefaults?) {
        self.defaults = defaults
    }

    public var favoriteTexts: Set<String> {
        Set(defaults?.stringArray(forKey: Self.key) ?? [])
    }

    public func isFavorite(_ quote: Quote) -> Bool {
        favoriteTexts.contains(quote.text)
    }

    public func toggleFavorite(_ quote: Quote) {
        var favorites = favoriteTexts
        if favorites.contains(quote.text) {
            favorites.remove(quote.text)
        } else {
            favorites.insert(quote.text)
        }
        defaults?.set(Array(favorites), forKey: Self.key)
    }
}

File: android/app/src/main/kotlin/com/coldopen/app/FavoritesStore.kt
package com.coldopen.app

import android.content.Context
import com.coldopen.core.Quote
import com.coldopen.app.widget.ColdOpenWidget

/**
 * Shared favorite-quote storage, backed by the same `SharedPreferences`
 * file [ColdOpenWidget]'s "already shown" persistence already uses —
 * written by the companion browser ([MainActivity]) when the user taps
 * the favorite toggle, read by the widget to exclude favorites from the
 * randomize/auto-refresh pool. Unlike iOS, this needs no App
 * Group-equivalent: the widget and the app run in the same process here.
 */
object FavoritesStore {
    private const val PREF_FAVORITE_TEXTS = "favorite_quote_texts"

    private fun prefs(context: Context) =
        context.getSharedPreferences(ColdOpenWidget.PREFS_NAME, Context.MODE_PRIVATE)

    fun favoriteTexts(context: Context): Set<String> =
        prefs(context).getStringSet(PREF_FAVORITE_TEXTS, emptySet()) ?: emptySet()

    fun isFavorite(context: Context, quote: Quote): Boolean =
        favoriteTexts(context).contains(quote.text)

    fun toggleFavorite(context: Context, quote: Quote) {
        val favorites = favoriteTexts(context).toMutableSet()
        if (!favorites.add(quote.text)) {
            favorites.remove(quote.text)
        }
        prefs(context).edit().putStringSet(PREF_FAVORITE_TEXTS, favorites).apply()
    }
}
